corehq/apps/cloudcare/static/cloudcare/js/README.md

```diff
@@ -0,0 +1,168 @@
+
+##### Regions
+
+Marionette's [regions](https://marionettejs.com/docs/master/marionette.region.html) are UI containers, defined in the FormplayerFrontend's `before:start` handler.
+
+We rarely touch the region-handling code, which defines the high-level structure of the page: the "main" region, the progress bar, breadcrumbs, and the restore as banner. The persistent case tile also has a region. Most web apps development happens within the `main` region.
+
+It is sometimes useful to know how the breadcrumbs work. The breadcrumbs are tightly tied to formplayer's selections-based navigation. See [Navigation and replaying of sessions](https://github.com/dimagi/commcare-hq/blob/master/docs/formplayer.rst#navigation-and-replaying-of-sessions) for an overview and examples. The breadcrumbs use this same selections array, which corresponds to the "steps" attribute of `CloudcareURL`, with one breadcrumb for each selection.
+
+##### Backbone.Radio and Events
+
+Marionette [integrates with Backbone.Radio](https://marionettejs.com/docs/master/backbone.radio.html) to support a global message bus.
+
+Although you can namespace channels, web apps uses a single `formplayer` channel for all messages, which is accessed using `FormplayerFrontend.getChannel()`. You'll see calls to get the channel and then call `request` to get at a variety of global-esque data, especially the current user. All of these requests are handled by `reply` callbacks defined in `FormplayerFrontend`.
+
+`FormplayerFrontend` also supports events, which behave similarly. Events are triggered directly on the `FormplayerFrontend` object, which defines `on` handlers. We tend to use events for navigation and do namespace some of them with `:`, leading to events like `menu:select`, `menu:query`, and `menu:show:detail`.
+
+Counterintuitively, `showError` and `showSuccess` are implemented differently: `showError` is an event and `showSuccess` is a channel request.
+
+##### Routing and Middleware
+
+Being a SPA, all of web apps' navigation is handled by a javascript router, `Marionette.AppRouter`, which extends Backbone's router. [router.js](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js) defines the web apps router.
+
+The router also handles actions that may not sound like traditional navigation in the sense that they don't change which screen the user is on. This includes actions like pagination or searching within a case list.
+
+Other code generally interacts with the router by triggering an event (see above for more on events). Most of `router.js` consists of event handlers that then call the router's API.
+
+Every function in the router's API applies the function defined in [middleware.js](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cloudcare/static/cloudcare/js/formplayer/middleware.js). This is rarely edited. It's where the "User navigated to..." console log messages come from.
```

Comment:
Is this how it functions as middleware?  Like, it has to be manually called to mediate each request?